src/purge.rs

```diff
@@ -71,6 +146,10 @@ where
             warn!("Unable to purge expired files: locked");
             return Ok(vec![]);
         }
+        if self.rewrite_marker_path.exists() {
+            // Otherwise all writes will be erased next reboot.
+            remove_marker(&self.rewrite_marker_path)?;
```

Comment:
Why it's safe to remove it? Or why it exists in the first place?

Reply:
The last `purge` fails to remove it, subsequent purge will never write to rewrite queue unless the marker is removed here. (it's guaranteed that only purge will write to rewrite queue).

Reply:
Oh, here I only consider the case that purge (aka rewrite) succeeds but fails to remove the marker. I'll need to do log file clean up here as well to make it work.

As for your question, a failed `purge` doesn't necessarily mean the whole engine is shutdown. User can still call purge over it, he just has to wait until the cleanup is finished here.

Reply:
How to know it's the case that "rewrite succeeds but fail to delete file" or "rewrite fails and fail to delete file"?

Reply:
We will always treat it as the latter. The code is updated to protect the entire purge operation as an atomic part.